[Skia] Implement support for shadows that ignore transforms

GeorgesStavracas opened the pull request:
#### caf39945f1d1355f2da6049b919218f0c2b68b7b
<pre>
[Skia] Implement support for shadows that ignore transforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=270175">https://bugs.webkit.org/show_bug.cgi?id=270175</a>

Reviewed by Carlos Garcia Campos and Nikolas Zimmermann.

When creating the drop shadow filter, if state.shadowsIgnoreTransforms()
is true, render the drop shadow with trasformed coordinates and blur
radius, by adjusting them against the CTM inverse.

This is enough to make various canvas tests happy. Some of them are still
failing with off-by-one colors, and it&apos;s not clear why.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::createDropShadowFilterIfNeeded const):

Canonical link: <a href="https://commits.webkit.org/275492@main">https://commits.webkit.org/275492@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e124da8d10ce644950da90cc175a648b6cd1c9f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42001 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44587 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38113 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44308 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24194 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18347 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42575 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/17951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36171 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37202 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46011 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38195 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37529 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13823 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/18436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5633 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/18081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -291,25 +292,46 @@ void GraphicsContextSkia::strokePath(const Path& path)
     ASSERT(shadow);
 
-    const FloatSize& offset = shadow->offset;
+    FloatSize offset = shadow->offset;
```
auto?

GeorgesStavracas replied:
Done

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -291,25 +292,46 @@ void GraphicsContextSkia::strokePath(const Path& path)
+
+            // Ignoring the CTM is practically equal as applying the inverse of
+            // the CTM when post-processing the drop shadow.
+            if (const std::optional<SkMatrix>& inverse = ctm.inverse()) {
```
Any particular reason for converting directly to SkMatrix? IMHO AffineTransform is more readable.

GeorgesStavracas replied:
It's because the methods `mapHomogeneousPoints` and `mapRadius` only exist in SkMatrix, unless I've missed something?

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -291,25 +292,46 @@ void GraphicsContextSkia::strokePath(const Path& path)
+            // the CTM when post-processing the drop shadow.
+            if (const std::optional<SkMatrix>& inverse = ctm.inverse()) {
+                SkPoint3 p = SkPoint3::Make(offset.width(), offset.height(), 0);
+                inverse->mapHomogeneousPoints(&p, &p, 1);
```
Makes sense! You need to transform all lengths to be in identity space - as you intent to remove the CTM.

nikolaszimmermann reviewed (approved):
r=me with comments

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -291,25 +292,46 @@ void GraphicsContextSkia::strokePath(const Path& path)
+        if (state.shadowsIgnoreTransforms()) {
+            // state.shadowsIgnoreTransforms() means the offset is in natural
+            // coordinates (Y is negative above the X line), convert that back
+            // to the Skia coordinate system
```
I think the reason for this is that CG uses natural orientation for Y axis.

GeorgesStavracas replied:
Thanks. Updated the comment.

GeorgesStavracas committed:
[Skia] Implement support for shadows that ignore transforms
https://bugs.webkit.org/show_bug.cgi?id=270175

Reviewed by Carlos Garcia Campos and Nikolas Zimmermann.

When creating the drop shadow filter, if state.shadowsIgnoreTransforms()
is true, render the drop shadow with trasformed coordinates and blur
radius, by adjusting them against the CTM inverse.

This is enough to make various canvas tests happy. Some of them are still
failing with off-by-one colors, and it's not clear why.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::createDropShadowFilterIfNeeded const):

Canonical link: https://commits.webkit.org/275492@main
